chore(ruff): configure ruff

piotr-roslaniec opened the pull request:
Required reviews: 1
Based on: https://github.com/nucypher/nucypher-ops/pull/32

- Adds CI boilerplate for building, lining (and in the future - testing)
- Adds Python versions matching `nucypher/nucypher`
- Adds `ruff` for linting, fixes lint errors

piotr-roslaniec committed:
chore(ruff): configure ruff

derekpierre commented on .github/workflows/python.yml:
```diff
@@ -0,0 +1,30 @@
+        python-version: [ "3.9", "3.10", "3.11", "3.12" ]
+
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v4
```

derekpierre commented on .github/workflows/python.yml:
```diff
@@ -0,0 +1,30 @@
+      - uses: actions/checkout@v2
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
```
```suggestion
        uses: actions/setup-python@v4
```

derekpierre commented on .github/workflows/python.yml:
```diff
@@ -0,0 +1,30 @@
+          pip install setuptools wheel ruff
+
+      - name: Run ruff
+        run: ruff check nucypher_ops
```
Looking at `nucypher`'s use of ruff (see https://github.com/nucypher/nucypher/blob/main/.github/workflows/ruff.yaml), should this be?
```
ruff --output-format=github nucypher_ops
```

Just wondering if there is a reason for the discrepancy.

piotr-roslaniec replied:
No reason, I don't know what `--output-format=github` does.

derekpierre replied:
I believe (?) the format caters to specific output that can be viewed as gh actions workflow outputs. Eg. https://github.com/nucypher/nucypher/actions/runs/6708762664/job/18230283150

derekpierre commented on .github/workflows/python.yml:
If this is only for ruff (and I'm assuming here) then should this file be called `ruff.yaml` instead?

piotr-roslaniec replied:
`python.yml` sounds fine to me, to be honest

derekpierre commented on .github/workflows/python.yml:
```diff
@@ -0,0 +1,30 @@
+on: [ push, pull_request ]
+
+jobs:
+  test:
```
Assuming this action is just for `ruff` let's name this `ruff` or `ruff-lint` or something analogous instead of `test`.

piotr-roslaniec replied:
`python.yml` sounds fine to me, to be honest

piotr-roslaniec commented on .github/workflows/python.yml:
```diff
@@ -0,0 +1,30 @@
+    runs-on: ubuntu-20.04
+    strategy:
+      matrix:
+        python-version: [ "3.9", "3.10", "3.11", "3.12" ]
```
It's not just to run ruff, but rather configure CI as a whole and add `ruff` linting on top of that. I added clarification to the issue description

piotr-roslaniec committed:
Update .github/workflows/python.yml

Co-authored-by: Derek Pierre <derek.pierre@gmail.com>

Update .github/workflows/python.yml

Co-authored-by: Derek Pierre <derek.pierre@gmail.com>

Update nucypher_ops/__about__.py

Co-authored-by: Derek Pierre <derek.pierre@gmail.com>

Update nucypher_ops/__about__.py

Co-authored-by: Derek Pierre <derek.pierre@gmail.com>

piotr-roslaniec committed:
apply pr suggestions

derekpierre commented on .github/workflows/python.yml:
```diff
@@ -0,0 +1,42 @@
+      - name: Build dist
+        run: python setup.py sdist bdist_wheel
+
+  lint:
```
👌

derekpierre commented on pyproject.toml:
```diff
@@ -0,0 +1,7 @@
+[tool.ruff]
+exclude = ["nucypher_ops/__init__.py"] # false positives
+select = ["E", "F", "I"]
+line-length = 180
```
Pretty long line length 😅 - although it's hard to specify a value that everyone agrees on hehe.

piotr-roslaniec replied:
I prefer shorter lines myself, but it would be very much out of style with the current code style in `nucypher-ops`

derekpierre replied:
👍. Another solution is to ignore line length errors (https://docs.astral.sh/ruff/rules/line-too-long/) i.e. add this to the [tool.ruff] section:
```
ignore = ["E501"]
```

then you can remove the `line-length` setting. Up to you though.

derekpierre reviewed (approved):
🎸

piotr-roslaniec committed:
chore: fix linter after rebase

derekpierre commented on .github/workflows/python.yml:
```diff
@@ -3,23 +3,43 @@ name: Python package CI
 
     steps:
       - uses: actions/checkout@v4
+<<<<<<< HEAD
+=======
+
+>>>>>>> cced0c2 (Update .github/workflows/python.yml)
```
@piotr-roslaniec did this get merged with conflicts?
